man1/ftff.1
.\"
.\"     This is free software and distributed under the
.\"     terms of the Gnu Public License.
.\"
.TH FTFF 1 "August 1998" "Search utilities"  \" -*- nroff -*-
.SH NAME
ftff \- fault tolerant file find utiltiy
.SH SYNOPSIS
.B ftff
[\-#fFhIpq][\-t#][start_directory] file_to_find
.SH DESCRIPTION
.B ftff
recursively descends the directory hierarchy and reports
all objects in the file system with a name that approximately
matches the given filename.
.PP
.B ftff
achieves  fault tolerance  by calculating the  so called
Weighted Levenshtein Distance. The Levenshtein Distance
is defined as the  minimum  number of  character insertions,
deletions and replacements that transform 
a string 
.I A 
into a string 
.I B.
.PP
.B ftff
behaves like 
.PP
.RB 'find\ start_directory\ \-name\ file_to_find\ \-print' 
.PP
with the following differences:
.TP
\- 
ftff is fault tolerant
.TP
\- 
ftff is NOT case sensitive
.TP
\- 
the level of fault tolerance can be adjusted by specifying the optional parameter 
.B tolerance.
A 
.B tolerance
of 0 specifies exact match.
.SH OPTIONS
.TP
.I \-h
Prints a little help/usage information.
.TP
.I \-f
Follow symbolic links on directories. Note: a symbolic link
like "somewhere \-> .."  causes naturally an endless loop. 
By default
.B ftff
does not follow symbolic links to directories.
.TP
.I \-F
Classify the file type by appending a character to each file name. This
character is:
.br
.I '*' 
for regular files that are  executable
.br
.I '/' 
for directories
.br
.I '@' 
for symbolic links
.br
.I '|' 
for FIFOs
.br
.I '=' 
for sockets
.TP
.I \-p
print the actual distance value in front of the filename. This value
is equal to the number of insertions, deletions and replacements necessary
to transform the file that was found into the search key (the file_to_find).
.TP
.I \-q
keep quiet and do not print any warning about non readable directories.
.TP
.I \-# or \-t#
Set the fault tolerance level to #.
The fault tolerance level is an integer in the range 0-255.
It specifies the maximum number of errors permitted in finding 
the approximate match. The default tolerance is (strlen(searchpattern) - number of wildcards)/6 + 1
.TP
.I \-I
Do case sensitive search (default is case in-sensitive)
.PP
.TP 
.I file_to_find
The filename to search for.  
.RB '*'
and 
.RB '?' 
can be used as wildcards.  
.br
.RB '?' 
denotes one single character.  
.br 
.RB '*' 
denotes an arbitrary number of characters.
.TP 
.I start_directory
The directory to start the search. The current directory is the default. 
.PP
The last argument to ftff is not parsed for options as the program needs
at least one file-name argument. This means that 
.I ftff -x
will not complain about a wrong option but search for the file named -x.
.SH EXAMPLE
.I  ftff 
.RB samething
.PP
This will e.g. find a file called something or sameting or sum-thing or ...
.PP
To find all files that start with any prefix, have something like IOComm in between and end on a two letter suffix:
.PP
.I  ftff 
.RB '*iocomm.??'
.PP
To find all files that exactly start with the prefix DuPeg:
.PP
.I ftff \-0 
.RB 'dupeg*' 
.PP
.SH BUGS
The wildcards 
.BR '?'
and 
.BR '*'
can not be escaped. These characters function always as
wildcards. This is however not a big problem since there is
normally hardly any file that has these characters in its name.
.SH AUTHOR
Guido Socher (guido@linuxfocus.org)
.SH SEE ALSO
whichman(1) find(1)



debian/changelog
whichman (2.4-8) unstable; urgency=medium

  * Add 11-CPPFLAGS-and-LDFLAGS.patch to pass the two variables
    to compiler.
  * Cleanup debian/rules, add support for cross-compiling, and
    enable LFS support.
  * Standards-Version: 3.9.5 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 14 Dec 2013 21:35:44 +0100

whichman (2.4-7) unstable; urgency=low

  * Switch to debhelper v9 and tiny rules file.
  * Rename & refresh patches with gbp-pq import/export.
  * Fix lintian's `spelling-error-in-copyright'.
  * debian/rules: remove .pc/.dpkg-source-unapply file in rules.
    This fixes broken behaviour of dpkg-buildpackage (see Bug#649521).
  * debian/control:
    + Standards-Version: 3.9.2 (no changes);
    + add VCS fields;
    + sort dependency fields with wrap-and-sort.

 -- Robert Luberda <robert@debian.org>  Sun, 19 Feb 2012 16:05:46 +0100

whichman (2.4-6) unstable; urgency=low

  * Switch to the `3.0 (quilt)' format.
  * Bump to debhelper v8.
  * Standards-Version: 3.9.1 (no changes).
  * Split the previous 01_debian.patch into 3 patches: 00_Makefile.patch,
    01_documentation.patch, and 02_no_warn.patch.
  * debian/rules:
    + use dpkg-buildflags for getting CFLAGS/LDFLAGS (also added versioned
      build-dependency on dpkg-dev for this);
    + call dh_auto_{configure,build,clean} commands;
    + provide build-{arch,indep} targets.

 -- Robert Luberda <robert@debian.org>  Tue, 15 Feb 2011 22:29:25 +0100

whichman (2.4-5) unstable; urgency=low

  * Add copyright notice (lintian),

 -- Robert Luberda <robert@debian.org>  Sat, 19 Apr 2008 16:04:29 +0200

whichman (2.4-4) unstable; urgency=low

  * Use quilt to manage patches, move old debian patch to 01_debian.patch.
  * Build with debhelper v6.
  * 10_skip_duplicates.patch: Avoid duplicated files from `ftwhich' and
    `whichman' output (closes: #452068).
  * Standards-Version: 3.7.3 (no changes).
  * debian/control: Add Homepage URL.

 -- Robert Luberda <robert@debian.org>  Tue, 29 Jan 2008 22:39:17 +0100

whichman (2.4-3) unstable; urgency=low

  * Fix the `debian-rules-ignores-make-clean-error'.
    and manpage-has-errors-from-man' lintian errors.
  * Bump debhelper compat version to 6.
  * Standards-Version: 3.7.2 (no changes).

 -- Robert Luberda <robert@debian.org>  Mon, 15 Oct 2007 21:37:39 +0200

whichman (2.4-2) unstable; urgency=low

  * Fix typos in ftwhich(1) and ftff(1) man pages.
    (Closes: #306727,#306728).
  * Add debian/watch file.

 -- Robert Luberda <robert@debian.org>  Fri, 29 Apr 2005 19:06:43 +0200

whichman (2.4-1) unstable; urgency=low

  * New upstream version.

 -- Robert Luberda <robert@debian.org>  Tue, 20 Apr 2004 19:48:32 +0200

whichman (2.3-2) unstable; urgency=low

  * Opps, man pages went to /usr/share/man/man1/man1. Fixed.

 -- Robert Luberda <robert@debian.org>  Sun, 28 Mar 2004 23:15:17 +0200

whichman (2.3-1) unstable; urgency=low

  * New upstream version.
  * Standards-Version: 3.6.1 (no changes).

 -- Robert Luberda <robert@debian.org>  Wed, 24 Mar 2004 19:39:44 +0100

whichman (2.2-1) unstable; urgency=low

  * New upstream version.
  * debian/control: remove ending dot from the first line of the description
    field (lintian).
  * debian/copyright: update URLs.
  * Standards-Version: 3.5.8 (no changes).

 -- Robert Luberda <robert@debian.org>  Thu,  9 Jan 2003 00:01:13 +0100

whichman (2.0-3) unstable; urgency=low

  * Upgrade to Standards: 3.5.7
  * Build in debhelper in v4 mode.

 -- Robert Luberda <robert@debian.org>  Tue, 22 Oct 2002 22:20:23 +0200

whichman (2.0-2) unstable; urgency=low

  * Corrected spelling error in description (closes: #125481) and upstream
    README file. Thanks to Matt Ziemmerman.
  * Standards-Version: 3.5.6 (no changes).

 -- Robert Luberda <robert@debian.org>  Sat, 22 Dec 2001 22:22:00 +0100

whichman (2.0-1) unstable; urgency=low

  * New upstream version.
  * Fixed gcc 3.0 warnings.

 -- Robert Luberda <robert@debian.org>  Fri,  6 Jul 2001 22:49:21 +0200

whichman (1.9-2) unstable; urgency=low

  * Some fixes in manpages.
  * Updated debian/rules to the newest debhelper.
  * Standards-Version: 3.5.5.

 -- Robert Luberda <robert@debian.org>  Wed,  4 Jul 2001 22:58:46 +0200

whichman (1.9-1) unstable; urgency=low

  * Initial release (closes: #83624).
  * Use default Debian manpath if environment variable MANPATH
    is not set.

 -- Robert Luberda <robert@debian.org>  Mon, 29 Jan 2001 20:30:47 +0100


ftff.c
/* vim: set sw=8 ts=8 si et: */
/*
 * ftff, fault tolerant file-find.
 * This program is distributed under the terms 
 * of the Gnu Public License (GPL). 
 */
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <dirent.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "levdist.h"

#ifndef S_IXUSR
#define S_IXUSR S_IEXEC
#endif
#ifndef S_IXGRP
#define S_IXGRP (S_IEXEC >> 3)
#endif
#ifndef S_IXOTH
#define S_IXOTH (S_IEXEC >> 6)
#endif
#ifndef S_IXUGO
#define S_IXUGO (S_IXUSR | S_IXGRP | S_IXOTH)
#endif

static void help();

/* global for efficiency reasons: */
static char *search;
static char *fullpath;
static int tolerance=-1; /*-1 indicates tolerance is not set*/
static int opt_p=0;
static int opt_q=0;
static int opt_F=0;
static int opt_I=0;
static int opt_f=0;



#define PATH_MAX_LEN 1000

char *allocpath(void)
{
	char *ptr;
	if ((ptr = (char *) malloc(PATH_MAX_LEN + 2)) == NULL) {
		perror("Error: malloc() failed");
		exit(1);
	}
	return ptr;
}
/* classify the file type an produce a charater like ls -F */
char filetype(struct stat buffer){
#ifdef linux
        /*umode_t mode;*/
        /*__mode_t mode;*/
        mode_t mode;  /* these days it's all mode_t */
#else /* linux */
#ifdef sun
        /* solaris: */
        mode_t mode;
#else /* sun */
        /* other unix os lands here: */
        /*unsigned long mode;*/
        mode_t mode;
#endif /* sun */
#endif /* linux */


        mode=buffer.st_mode;
        switch(mode & S_IFMT) {
#ifdef S_IFIFO
        case S_IFIFO:
                return '|';
#endif
#ifdef S_IFDIR
        case S_IFDIR:
                return '/';
#endif
#ifdef S_ISLNK
        case S_IFLNK:
                return '@';
#endif
#ifdef S_IFSOCK
        case S_IFSOCK:
                return '=';
#endif
        case S_IFREG:
                if (mode & S_IXUGO) {
                        return '*';
                }
                return ' ';
        /* no default */
        }
        return ' ';
}

/*
 *check for the file names ".." and "."
 *return 0 if dir is exactly ".." or "." otherwise return 1
 */ 
int isnotdotdir(char *dir)
{
        int cnt=0;
        while(*dir){
                if(*dir != '.'){
                        return(1);
                }
                if (cnt > 1){
                        return(1);
                }
                cnt++;
                dir++;
        }
        return(0);
}

/* Traverse directory tree  but we don't follow symbolic links
 * unless opt_f is set. The link-names are processed in any case.
 *
 * Return 0 on sucessful file find.
 * 
 * object = "name of object in FS"
 * fullpath = "path"+object  (this is a global variable)
 * This is a recursive function.
 * Uses also the following global var: opt_p,opt_f,opt_I,opt_q
 */

int traversetree(char *object)
{
	/* for efficiency reasons: */
	static struct stat buffer;
        /*these variables must be local:*/
	DIR *dp;
	struct dirent *dirp;
	char *ptr;
        /* must be static:*/
        static int result =1;

        int d,len;

        if (lstat(fullpath, &buffer) < 0) {
                if (opt_q==0){
                        fprintf(stderr, "Warning: %s, ", fullpath);
                        perror("");
                }
                return(result);
        }
#ifdef DEBUG
	fprintf(stderr, "debug,f:%s o:%s s:%s\n", fullpath,object,search);
#endif
        /*the first time that we are called obj is ".", skip that:*/
        if (isnotdotdir(object)){
                d=levdist(object, search, tolerance,opt_I);
                if (d != -1) {
                        if (opt_p) printf("%03d ",d);
                        printf("%s", fullpath);
                        if (opt_F){
                                printf("%c\n",filetype(buffer));
                        }else{
                                printf("\n");
                        }
                        result = 0;
                }
        }
        if (opt_f && S_ISLNK(buffer.st_mode)){
                if (stat(fullpath, &buffer) < 0) {
                        if (opt_q==0){
                                fprintf(stderr, "Warning: %s, ", fullpath);
                                perror("");
                        }
                        return(result);
                }
        }
	if (S_ISDIR(buffer.st_mode)) {
		if (NULL == (dp = opendir(fullpath))) {
			if (opt_q==0){
                                fprintf(stderr, "Warning: %s, ", fullpath);
                                perror("");
                        }
			return(result);
		}
                len=strlen(fullpath);
		ptr = fullpath + len;
                /* protection against endless loops due to symbolic links
                 * to .. */
                if (len < PATH_MAX_LEN -100){
                        if (*(ptr -1) != '/'){
                                /* do not append a "/" if the user did already 
                                 * provide one in the start_dirictory name 
                                 */
                                *ptr = '/';
                                ptr++;
                        }
                        *ptr = '\0';
                        while (NULL != (dirp = readdir(dp))) {
                                if (isnotdotdir(dirp->d_name)){
                                        /* Append to fullpath without changing 
                                         * ptr this will always succeed since 
                                         * we allocated enough memory.*/
                                        strcpy(ptr, dirp->d_name);
                                        traversetree(dirp->d_name);
                                }
                        }
                        closedir(dp);
                        /*shorten fullpath again to current directory:*/
                        ptr--;
                }
                *ptr='\0';
	}
        return (result);
}


int main(int argc,char *argv[])
{
        int result,optnum;
	struct stat buf;
	char *c,*oarg;

        if (argc < 2 ) help();
        /*
         * After this while loop. argv[0] is the next argument if available
         *
         * options that take an argument write *c=0;break; others write
         * "c++;break;"
         *
         * This parser looks a bit complicated but it can take options
         * like -23 and -ac is the same as "-a -c" Because we number options
         * are allowed we can not use on of the standard option parsers.
         */
        optnum=argc;
        while (--optnum>0) {
                argv++;
                c = *argv; /* c is pointer to one argument/option */
                if (*c != '-'){
                        argv--;
                        break; /*stop at the first non option argument*/
                }
                c++;
                if (*c == '-' && *(c + 1)=='\0'){
                        /* you can write a double -- to stop option
                         * parsing. This is useful if you need to search
                         * for a string that is called "-something" */
                         --optnum;
                         break; /*stop option parsing*/
                }
                while (*c) {
                        switch (*c) {
                        case 'f':
                                opt_f=1;
                                c++;break;
                        case 'F':
                                opt_F=1;
                                c++;break;
                        case 'h':
                                help();
                                /*no break, help does not return*/
                        case 'p':
                                opt_p=1;
                                c++;break;
                        case 'q':
                                opt_q=1;
                                c++;break;
                        case 'I':
                                opt_I=1;
                                c++;break;
                        case 't':
                                /* you can write -t11 or -t 11*/
                                /* first the normal -t11 case: */
                                oarg=c+1;
                                /* then the -t 11 case: */
                                if(*(c + 1)=='\0'){
                                        if(--optnum) oarg=*++argv;
                                }
                                tolerance = atoi(oarg);
                                if (tolerance > 255 || tolerance <0) tolerance = 255;
                                *c=0;break;
                        default:
                                if (isdigit((int)*c)){
                                        tolerance = atoi(c);
                                        if (tolerance > 255 || tolerance <0) tolerance = 255;
                                }else{
                                        fprintf(stderr,"ERROR: No such option. Use -h to get help.\n");
                                        exit(1);
                                }
                                *c=0;
                        }
                }
        }
        if (optnum<0) optnum=0; /* can be less then zero due to a option
                                 * with arg at last pos and missing arg*/
        /*
         *optnum  is now the number of argmunts left after
         *option parsing. Example: option a,c take no arg, option b has arg:
         *                         thisprogram -a -b xx -c d e
         *                         This results in:
         *                         optnum==2 and argv[1] points to d
         *                         argv[2] points to e
         */ 
        fullpath = allocpath();
	if (1 == optnum) {
		search = argv[1];
                /*start directory is current directory*/
                strcpy(fullpath, ".");
	} else if (2 == optnum) {
		search = argv[2];
                strncpy(fullpath, argv[1],PATH_MAX_LEN);
                fullpath[PATH_MAX_LEN]=0;
                if (stat(fullpath, &buf) < 0) {
                        fprintf(stderr, "ERROR: no such directory, %s\n", fullpath);
                        exit(1);
                }
                if (!S_ISDIR(buf.st_mode)){
                       fprintf(stderr, "ERROR: %s is not a directory -h for help.\n",fullpath);
                        exit(1);
                }
	} else {
                help();
	}
        if (tolerance == -1) tolerance=stdtolerance(search);
	result=traversetree(".");
	return result;
}
/*
 * help
 */
void help()
{
printf("ftff -- fault tolerant file find utility\n\
\n\
USAGE: ftff [-#fFhIpq][-t#] [--] [start_dir] filename \n\
\n\
ftff uses an approximate string matching algorithm to match the file names.\n\
\n\
Supported wildcards: * : any arbitrary number of character\n\
                     ? : one character\n\
\n\
OPTIONS: -f  follow symbolic links on directories.\n\
         -F  classify the file type like \"ls -F\".\n\
         -h  this help.\n\
         -I  search case sensitive (default is case in-sensitive)\n\
         -p  print the actual distance value in front of the file name.\n\
         -q  Quiet. Do not print a warning when a directory can not be read.\n\
         -#  set fault tolerance level to # (integer in the range 0-255) \n\
             This is the maximum number of permitted errors.\n\
         -t# same as -# for backward compatibility\n\
\n\
The default start_dir is the current directory. The default fault tolerance\n\
is (string length of searchpattern - number of wildcards)/6 +1.\n");
printf("%s\n",MYVERSION);
exit(0);
}
/*__END__*/
